[CORE-434] Don't filter out resources marked as deleted

samanehsan opened the pull request:
<!-- Welcome to your Leonardo pull request! -->
<!-- Contributor guidelines: https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md -->
Ticket: https://broadworkbench.atlassian.net/browse/CORE-434

## Summary of changes

<!-- Please give an abridged version of the ticket description here and/or fill out the following fields. -->

### What

Leo may think a resource was deleted in Sam when it actually still exists and then skip making the delete call to Sam. When a user goes to delete the workspace, it will fail because the child resource record still exists. The solution is to not filter out those resources so that they are deleted in Sam. 

The other change we discussed was: 
> Whenever Leo marks a resource as deleted in its own database, it should also attempt to delete the Sam resource. This will prevent future workspaces from getting into this undeletable state.

But it looks like this is already happening:
- [Runtimes](https://github.com/DataBiosphere/leonardo/blob/68e204738a01ba34b78f673098635e4c4ef111d9/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeServiceInterp.scala#L393) 
- [Apps](https://github.com/DataBiosphere/leonardo/blob/68e204738a01ba34b78f673098635e4c4ef111d9/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala#L517)
- [Disks](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/DiskServiceInterp.scala#L310)

### Why

This fixes a bug that is preventing users from deleting workspaces.

## Testing these changes

### What to test

- Test that even if a runtime is marked as deleted in Leo, its Sam record gets deleted (if it still exists)

<!-- ### Test data -->

### Who tested and where
- Tested in a BEE
- Steps:
  - Create a workspace with a notebook
  - Look up the notebook record in Sam
  - Connect to the Leo database and change the runtime db record `status` to `Deleted`
  - Call `/api/google/v1/resources/{googleProject}/cleanupAll` and confirm that the notebook record in Sam was deleted

- [x] This change is covered by automated tests <!-- (unit, automation, contract, etc) -->
  - _NB: Rerun automation tests on this PR by commenting `jenkins retest` or `jenkins multi-test`._
- [x] I validated this change <!-- (in what environment?) -->
- [ ] Primary reviewer validated this change <!-- (consider a pair review!) -->
- [ ] I validated this change in the dev environment <!-- (after successfully merging to `develop`) -->

samanehsan committed:
Don't filter out resources marked as deleted

Leo may think a resource was deleted in Sam
when it actually still exists, so do not filter
out those resources.

lucymcnatt commented:
Just wondering, why is the target specifically the listRuntimes endpoint? You may need to alter the other list endpoints as well (listDisks, listApps) to fully capture all potentially not-fully-deleted resources

samanehsan commented:
I think `listRuntimes` is the only one filtering out resources marked as deleted in Leo but I'll double check!

davidangb commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeServiceInterp.scala:
```diff
@@ -250,13 +250,11 @@ class RuntimeServiceInterp[F[_]: Parallel](
       _ <- ctx.span.traverse(s => F.delay(s.addAnnotation("Start DB query for listRuntimes")))
       runtimes <- RuntimeServiceDbQueries
         .listRuntimes(samResources.map(RuntimeSamResourceId).toSet,
-                      excludeStatuses = excludeStatuses,
```
I commented this in Slack, but figured I'd put this question in the PR: where else is this used? Will this start showing  deleted runtimes in Terra UI or from public-facing APIs?

LizBaldo replied:
I am a bit hesitant to include deleted resources here, because we should be relying on Sam about whether the user has the proper authorization to view the resource or not, which we cannot do for deleted resources since they are not supposed to exist in Sam anymore.

Making sure that this change does not show deleted runtimes in either the UI or swagger is definitely something to test (I think I added a unit test to make sure of that).

I think in this case, we only want to not filter out by deleted status for the deletion call, not the list one.

samanehsan committed:
Fix excluding deleted runtimes

samanehsan committed:
Fix syntax error

samanehsan committed:
Scala fmt

samanehsan committed:
Fix tests

samanehsan committed:
Exclude deleted resources from list APIs

samanehsan commented:
That's a good point @LizBaldo! I updated the list endpoints so that they still exclude deleted records. 

The reason I changed the list runtimes method is because `deleteAllResourcesRecords` calls `deleteAllRuntimesRecords` https://github.com/DataBiosphere/leonardo/blob/68e204738a01ba34b78f673098635e4c4ef111d9/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeServiceInterp.scala#L400 

which lists all the runtimes (excluding what it marked as deleted) and then deletes those runtime records from Sam. Some of the runtimes that Leo deleted still have records in Sam though so we also want to include those here. It is not clear to me how it gets into this state in the first place though. Maybe it's just that the call to Sam can fail sometimes (despite the retries)?

samanehsan committed:
Exclude deleted runtimes again

LizBaldo commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeService.scala:
```diff
@@ -34,6 +34,14 @@ trait RuntimeService[F[_]] {
     as: Ask[F, AppContext]
   ): F[Vector[ListRuntimeResponse2]]
 
+  def listRuntimes(userInfo: UserInfo,
```
Why create a new listRuntime function instead of modifying the previous one?

samanehsan replied:
It is called a bunch of times in `RuntimeServiceInterpSpec.scala` so I thought it would be nice to optionally accept `excludeStatuses: List[RuntimeStatus]` and make it default to excluding deleted runtimes (the current behavior) when excludeStatuses is not passed in. If you think it's clearer to just have the one function I'm happy to change it though!

LizBaldo replied:
Bleh, okay, it's not you, it's scala, I wish there was a clearer way to set this field as optional :(

lucymcnatt replied:
hmm could you not just add the following to the original function?
`excludeStatuses: List[RuntimeStatus] = List(RuntimeStatus.Deleted)`

that would set a default of excluding deleted runtimes

samanehsan commented:
The tests are passing now!

LizBaldo commented:
Nice, and were you able to delete a workspace with orphaned Sam resources in your BEE with this change?

samanehsan commented:
@LizBaldo I haven't tried to recreate/test this in a BEE but will definitely do that! To create an orphaned record do you think I can create a workspace with a runtime, delete the runtime in Leo and then add a child record back to Sam?

LizBaldo commented:
> @LizBaldo I haven't tried to recreate/test this in a BEE but will definitely do that! To create an orphaned record do you think I can create a workspace with a runtime, delete the runtime in Leo and then add a child record back to Sam?

That should do it yes :)

davidangb commented on http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/MockRuntimeServiceInterp.scala:
```diff
@@ -27,6 +27,15 @@ class BaseMockRuntimeServiceInterp extends RuntimeService[IO] {
   override def listRuntimes(userInfo: UserInfo, cloudContext: Option[CloudContext], params: Map[String, String])(
     implicit as: Ask[IO, AppContext]
+  ): IO[Vector[ListRuntimeResponse2]] =
+    listRuntimes(userInfo, cloudContext, params, List.empty)
```
I'm not sure how much it matters, but the logic here in the mock is not consistent with the logic in the real `RuntimeServiceInterp`. In `RuntimeServiceInterp`, this uses:
```scala
    val excludeStatuses = List(RuntimeStatus.Deleted)
    listRuntimes(userInfo, cloudContext, params, excludeStatuses)
```

davidangb commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/RuntimeServiceInterp.scala:
```diff
@@ -397,7 +407,7 @@ class RuntimeServiceInterp[F[_]: Parallel](
   ): F[Unit] =
     for {
-      runtimes <- listRuntimes(userInfo, Some(cloudContext), Map.empty)
+      runtimes <- listRuntimes(userInfo, Some(cloudContext), Map.empty, List(RuntimeStatus.Deleted))
```
since you have the two different signatures for `listRuntimes`, this could remain unchanged, right? It would default to `List(RuntimeStatus.Deleted)`?
```scala
runtimes <- listRuntimes(userInfo, Some(cloudContext), Map.empty)
```

samanehsan replied:
Oh, this needs to be `List.empty` because we do not want to exclude any resources (this would be the functional change). Thanks for catching that!

davidangb reviewed (commented):
syntax looks good. I think I'm missing something, though; quite possibly my lack of understanding of Leo. Where is the functional change in the PR that makes the difference? I might need a walkthrough …

samanehsan committed:
Fix excluded runtime statuses

Exclude deleted runtimes by default,
and exclude nothing when listing out
runtime records to delete.

davidangb reviewed (approved):
:+1: for the code changes; I leave it to you to test and verify it does what we need it to do!

LizBaldo reviewed (commented):
This approach makes sense to me for runtimes, thanks for checking on your BEE as well. I think you still need to add similar logic for apps ans disks too as these resources could also block workspace deletion if orphaned in sam

samanehsan commented:
Thanks for the review @LizBaldo! Where would you add that logic? When Leo gets the lists of apps and disks to delete in `deleteAllResourcesRecords`, it doesn't filter out anything that was marked as deleted like it was doing for runtimes.

LizBaldo commented:
Oh interesting, so listApps and listDisks would be returning `deleted` resources as well 🤔. If your BEE is still up, could you check that having an App/Disk marked as deleted in the leo DB (but not in Sam) is not preventing the workspace from deleting?

samanehsan committed:
Handle apps and persistent disks

samanehsan committed:
Fix GH actions

samanehsan commented:
@LizBaldo I checked in my BEE and any apps/disks with a `Deleted` status in Leo do not get listed. Deleted resources get filtered out by default. It looks slightly different from how runtimes does it and there is already an optional to include/exclude deleted resources, so I changed `deleteAllAppsRecords` and `deleteAllDisksRecords` to include deleted resources so that it ensures everything gets cleaned up.

LizBaldo reviewed (approved):
Thank you for doing that, It is nice to have that logic laid out in a similar way for all three resources. :shipit:

samanehsan committed:
Fix deleting app records

Always delete Sam records

samanehsan committed:
Merge branch 'develop' of github.com:DataBiosphere/leonardo into se/CORE-434-fix-resource-deletion-bug

samanehsan committed:
Fix syntax issues

davidangb commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala:
```diff
@@ -509,16 +507,26 @@ final class LeoAppServiceInterp[F[_]: Parallel](config: AppServiceConfig,
-      // Stop the usage of the SAS app
-      _ <- appUsageQuery.recordStop(dbApp.app.id, ctx.now).recoverWith { case e: FailToRecordStoptime =>
-        log.error(ctx.loggingCtx)(e.getMessage)
+      // check if the app is active
+      dbAppActiveOpt <- KubernetesServiceDbQueries
+        .getActiveFullAppByName(cloudContext, appName)
+        .transaction
+      _ <- dbAppActiveOpt match {
+        case Some(dbAppActive) =>
+          for {
```
you have already retrieved the `App` at line 490, above. Instead of re-querying the db, can you just match on that app's `status` and look for `AppStatus.Deleted`?

I see that `KubernetesServiceDbQueries.getActiveFullAppByName` does additional work related to clusters/cloud context, so I don't know if these are truly equivalent.

samanehsan replied:
The AppRecord has the `id` and `nodepoolId`, but not the cluster id. Instead of `KubernetesServiceDbQueries
        .getActiveFullAppByName(cloudContext, appName)` I could call `kubernetesClusterQuery.getMinimalActiveClusterByCloudContext` to just get the cluster id. Or update the query to get the app record to also include the cluster info.

davidangb replied:
ah, so they are actually different and required - no worries, then!

lucymcnatt replied:
nit, but it might be helpful if you added that context in a comment in this area, I agree with the reasoning though! You could probably add a `getFullAppByName` to `KubernetesServiceDbQueries` to grab everything in one query, but I'm also fine with the current solution

davidangb reviewed (approved):
one question inline, otherwise :+1:. I did not test this myself so I am trusting this solves the issue!

samanehsan commented:
I also realized I need to handle an error for deleting orphaned persistent disks...making that change now!

lucymcnatt commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/DiskServiceInterp.scala:
```diff
@@ -318,7 +318,7 @@ class DiskServiceInterp[F[_]: Parallel](config: PersistentDiskConfig,
-        Map.empty
+        Map(includeDeletedKey -> "true")
       )
       _ <- disks.traverse(disk => deleteDiskRecords(userInfo, cloudContext, disk))
```
Just thinking out loud, in deleteDiskRecords, there's a call to `persistentDiskQuery.delete`, which would be redundant if the Leo record is already set to 'deleted' and would overwrite the original `destroyedDate`. Not a huge deal, but maybe something to test to make sure that's working correctly in the case that Leo has already "deleted" the resource

samanehsan committed:
Delete orphaned disk records

samanehsan committed:
Handle errors when cleaning up deleted apps/disks

If a resource was marked as deleted in Leo, but there is
an orphaned Sam record this will delete the Sam record. If
there is no Sam record (because the resource was correctly
deleted the first time) there will be a Sam permissions error.

So if a resource is marked as deleted, just log the error and continue
deleting/cleaning up the remaining resources.

samanehsan committed:
Fix tests

samanehsan committed:
Set default excludeStatuses in listRuntimes

Remove the separate method that passes in a
default value.

samanehsan committed:
Add tests for app and disk cleanup

samanehsan committed:
Merge branch 'develop' of github.com:DataBiosphere/leonardo into se/CORE-434-fix-resource-deletion-bug

samanehsan committed:
Add getFullAppByName to get app info in one query

samanehsan committed:
Increase test coverage

lucymcnatt commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/LeoAppServiceInterp.scala:
```diff
@@ -528,9 +534,19 @@ final class LeoAppServiceInterp[F[_]: Parallel](config: AppServiceConfig,
+        Map(includeDeletedKey -> "true")
+      )
+      _ <- apps.traverse(app =>
+        deleteAppRecords(userInfo, cloudContext, app.appName).handleErrorWith { err =>
```
just to clarify, this is catching the case where Sam returns an error when checking permission for an app that doesn't exist?

samanehsan replied:
Yeah!

lucymcnatt reviewed (approved):
Thanks for addressing all our comments! I left one question but other than that LGTM

samanehsan committed:
Merge branch 'develop' into se/CORE-434-fix-resource-deletion-bug
